File: php/src/Engine/KeyGenerator.php
<?php

declare(strict_types=1);

namespace RespBench\Engine;

use RespBench\Config\KeyspaceConfig;

/**
 * Key generator for benchmark operations.
 *
 * Produces identical key sequences to the Java reference implementation (and the
 * Ruby/Python/Node engines) for the same seed.
 *
 * Two algorithms:
 *  - sequential_int: keys 0, 1, 2, ... N-1 (wraps around)
 *  - uniform_rand:   random keys using the Java-compatible LCG (JavaRandom)
 *
 * CROSS-PROCESS PARTITIONING: In the PHP engine each connection runs in its own
 * forked process, so a generator instance is never shared across processes and a
 * naive per-worker counter starting at 0 would make every worker write the SAME
 * keys — populating only `keys_count / N` distinct keys. To reproduce Java's key
 * set (whose forked threads share one counter and together walk 0..total-1), a
 * `sequential_int` worker strides its own disjoint subset: worker `i` of `N`
 * emits indices i, i+N, i+2N, … The union over all workers is exactly the full
 * keyspace with no duplicates (the interleaving order differs from Java's, which
 * does not matter for which keys end up populated).
 */
final class KeyGenerator
{
    private readonly string $keyPrefix;
    private readonly int $keySizeBytes;
    private readonly int $keysCount;
    private readonly int $seed;
    private readonly bool $sequential;
    private readonly int $workerIndex;
    private readonly int $workerCount;

    private int $sequentialStep = 0;
    private JavaRandom $random;

    public function __construct(
        private readonly KeyspaceConfig $config,
        ?int $seedOverride = null,
        int $workerIndex = 0,
        int $workerCount = 1,
    ) {
        $this->keyPrefix = $config->effectiveKeyPrefix();
        $this->keySizeBytes = $config->keySizeBytes;
        $this->keysCount = $config->keysCount;
        $this->seed = $seedOverride ?? $config->seedValue();
        $this->sequential = $config->isSequentialInt();
        $this->workerIndex = max(0, $workerIndex);
        $this->workerCount = max(1, $workerCount);
        $this->random = new JavaRandom($this->seed);
    }

    public static function create(KeyspaceConfig $config): self
    {
        return new self($config);
    }

    public static function createWithSeed(KeyspaceConfig $config, int $seed): self
    {
        return new self($config, $seed);
    }

    /**
     * Create a generator for worker `workerIndex` of `workerCount`.
     *
     * - sequential_int: strides a disjoint subset (i, i+N, i+2N, …).
     * - uniform_rand:   seeds per worker (seed + workerIndex) for reproducible,
     *   distinct random streams.
     */
    public static function forWorker(KeyspaceConfig $config, int $workerIndex, int $workerCount): self
    {
        $seedOverride = $config->isUniformRand()
            ? $config->seedValue() + $workerIndex
            : null;

        return new self($config, $seedOverride, $workerIndex, $workerCount);
    }

    /**
     * Generate the next key.
     */
    public function nextKey(): string
    {
        if ($this->sequential) {
            // Stride: worker i emits i, i+N, i+2N, … (mod keysCount).
            $keyIndex = ($this->workerIndex + $this->sequentialStep * $this->workerCount) % $this->keysCount;
            $this->sequentialStep++;
        } else {
            $keyIndex = $this->random->nextInt($this->keysCount);
        }

        $keyIndex %= $this->keysCount;

        return $this->formatKey($keyIndex);
    }

    /**
     * Reset the generator to its initial state.
     */
    public function reset(): void
    {
        $this->sequentialStep = 0;
        $this->random->setSeed($this->seed);
    }

    /**
     * Format a key index into a full key string.
     * Matches Java's String.format("%0Nd", keyIndex): zero-padded so that
     * prefix + number is approximately key_size_bytes wide (minimum 1 digit).
     */
    private function formatKey(int $keyIndex): string
    {
        $paddingWidth = max($this->keySizeBytes - strlen($this->keyPrefix), 1);

        return $this->keyPrefix . sprintf('%0' . $paddingWidth . 'd', $keyIndex);
    }
}

File: go/internal/engine/keygenerator.go
package engine

import (
	"fmt"
	"sync"
	"sync/atomic"

	"github.com/resp-bench/go/internal/config"
)

// KeyGenerator produces keys identical to the reference engines.
//
// Two algorithms:
//   - sequential_int: indices 0,1,2,… from a counter SHARED across all workers,
//     wrapping at keys_count. A shared atomic counter guarantees the union of
//     all workers covers the keyspace exactly once per pass with no duplication.
//   - uniform_rand: a PER-WORKER JavaRandom seeded base_seed+worker_index, so
//     each worker is independent yet reproducible.
//
// Keys are zero-padded to key_size_bytes: "%0Nd" where N = max(1, key_size_bytes
// - len(prefix)), matching Java's formatting.
type KeyGenerator struct {
	prefix       string
	keySizeBytes int
	keysCount    int
	sequential   bool

	// sequential_int: shared counter (nil for uniform_rand generators).
	counter *atomic.Int64

	// uniform_rand: per-worker PRNG (nil for sequential generators).
	rng *JavaRandom
	mu  sync.Mutex
}

// KeyGenSet holds the shared state for a phase's key generation so that all
// per-worker generators created via ForWorker share one sequential counter.
type KeyGenSet struct {
	cfg     config.KeyspaceConfig
	counter *atomic.Int64
}

// NewKeyGenSet builds the shared key-generation state for a phase.
func NewKeyGenSet(cfg config.KeyspaceConfig) *KeyGenSet {
	return &KeyGenSet{cfg: cfg, counter: &atomic.Int64{}}
}

// ForWorker returns a KeyGenerator for a specific worker. Sequential generators
// share the set's atomic counter; random generators get their own PRNG seeded
// base_seed + workerIndex.
func (s *KeyGenSet) ForWorker(workerIndex int) *KeyGenerator {
	kg := &KeyGenerator{
		prefix:       s.cfg.EffectiveKeyPrefix(),
		keySizeBytes: s.cfg.KeySizeBytes,
		keysCount:    s.cfg.KeysCount,
		sequential:   s.cfg.IsSequentialInt(),
	}
	if kg.sequential {
		kg.counter = s.counter
	} else {
		kg.rng = NewJavaRandom(s.cfg.SeedValue() + int64(workerIndex))
	}
	return kg
}

// keysCountOrOne guards against a zero keys_count causing a divide-by-zero.
func (k *KeyGenerator) keysCountOrOne() int {
	if k.keysCount <= 0 {
		return 1
	}
	return k.keysCount
}

// NextKey returns the next key as a string.
func (k *KeyGenerator) NextKey() string {
	var idx int
	if k.sequential {
		// fetch-and-increment, then wrap.
		n := k.counter.Add(1) - 1
		idx = int(n % int64(k.keysCountOrOne()))
	} else {
		k.mu.Lock()
		idx = k.rng.NextInt(k.keysCountOrOne())
		k.mu.Unlock()
	}
	return k.formatKey(idx)
}

// formatKey zero-pads the index to fill key_size_bytes after the prefix.
func (k *KeyGenerator) formatKey(idx int) string {
	pad := k.keySizeBytes - len(k.prefix)
	if pad < 1 {
		pad = 1
	}
	return fmt.Sprintf("%s%0*d", k.prefix, pad, idx)
}
